### DISCUSSION TITLE
Improve error messages

### DISCUSSION THREAD

**Discussion by cmi23** [wevm/viem#581] [Idea / Feature Request]
`ContractFunctionExecutionError: Execution reverted for an unknown reason.` is vague. I suggest returning more details, similar to ethers's approach.

`Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (reason="execution reverted: There are no open orders for token", method="estimateGas", transaction={"from":"0x8adDCAb8dC4Db2A74122EdD1ab3eF2197674DC4f","gasPrice":{"type":"BigNumber","hex":"0xd29b67c7"},"to":"0x74375D90b03981E2f3549C1e338BD0104fE414d2","value":{"type":"BigNumber","hex":"0x01"},"data":"0xfa4ef2fd000000000000000000000000acd7f3b07a3614df3eb2ce03d4ab446ad49418a00000000000000000000000000000000000000000000000000000000000000611","accessList":null}, error={"code":-32603,"data":{"code":3,"message":"execution reverted: There are no open orders for token","data":"0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000022546865726520617265206e6f206f70656e206f726465727320666f7220746f6b656e000000000000000000000000000000000000000000000000000000000000"}}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.7.2)`

**Comment by jxom** [comment]
We do return the [revert reasons](https://github.com/wagmi-dev/viem/blob/main/src/actions/public/estimateContractGas.test.ts#L189). Can you provide a minimal reproducible example?

**Reply by Surya-Git-hub** [reply]
hi jxom, i am facing the same issue in my project when calling useContractWrite and usePrepareContractWrite function.Not able to find the cause of error and error is send before writing transaction most probably because of usePrepareContract. minimal reproducible example is this on wagmi site examples >> https://wagmi.sh/examples/contract-write-dynamic

this one is on site
![Screenshot from 2023-06-14 00-35-37](https://github.com/wagmi-dev/viem/assets/64036341/290b9bb3-cbb4-4418-8d6b-3c36832778e0)

this one is of my project 
![Screenshot from 2023-06-14 00-37-35](https://github.com/wagmi-dev/viem/assets/64036341/6cb4f54c-5d1e-41d8-bc2b-cc90c10d465a)

**Comment by dev-aly3n** [comment]
facing same issue

**Reply by Surya-Git-hub** [reply]
i think if smart contract don't revert with reason, then viem shows unknown reason, like this solidity code >> " require(false);  "reverted without reason, where as if reason is present >> "  require(false," uncallable function "); " then viem will provide  reason as "uncallable function"

**Comment by arewageek** [comment]
How did y'all solve this?
Just ran into same issues while trying to write into my contract.

**Reply by tmm** [reply]
https://github.com/wevm/viem/discussions/581#discussioncomment-5999904

**Reply by micwallace** [reply]
@arewageek Did you try `e.walk()` on the original error? That worked for me.

**Reply by PaulRBerg** [reply]
Thanks @micwallace for your suggestion — I was able to obtain the revert data by using `.walk()`.

The Viem docs don't mention this feature .. I've opened an issue here: https://github.com/wevm/viem/issues/3520

**Comment by PaulRBerg** [comment]
Related:

- [Include revert data in the contract execution error types](https://github.com/wevm/viem/discussions/3519)
- [No documentation about how to obtain the revert data](https://github.com/wevm/viem/issues/3520)
